src/wtf8/mod.rs

```diff
@@ -23,17 +23,45 @@ pub struct Wtf8 {
 }
 
 impl Wtf8 {
+    /// Only safe if `bytes` contains valid WTF-8.
+    #[inline]
+    pub(crate) unsafe fn from_bytes_unchecked(bytes: &[u8]) -> &Self {
+        // Safety: the cast is sound because repr(transparent), matching the layout of [u8].
+        // As long as `bytes` is valid WTF-8, the type invariant is upheld.
+        &*(bytes as *const [u8] as *const Wtf8)
+    }
+    
+    /// Only safe if `bytes` contains valid WTF-8.
+    #[inline]
+    pub(crate) unsafe fn from_bytes_unchecked_mut(bytes: &mut [u8]) -> &mut Self {
+        // Safety: the cast is sound because repr(transparent), matching the layout of [u8].
+        // As long as `bytes` is valid WTF-8, the type invariant is upheld.
+        // The pointer is guaranteed to be safe for writes, as it comes from a mut ref.
+        &mut *(bytes as *mut [u8] as *mut Wtf8)
+    }
+    
     #[inline]
     pub(crate) fn bytes(&self) -> &[u8] {
         &self.bytes
     }
+    
+    #[inline]
+    pub(crate) fn bytes_mut<'a>(&'a mut self) -> &'a mut [u8] {
+        &mut self.bytes
+    }
 
     /// Coerces into a `Wtf8`. This accepts an [`&str`](prim@str) argument.
     #[inline]
     pub fn new<T: ?Sized + AsRef<Wtf8>>(x: &T) -> &Self {
         x.as_ref()
     }
 
+    /// Coerces into a mutable `Wtf8`. This accepts an [`&str`](prim@str) argument.
+    #[inline]
+    pub fn new_mut<T: ?Sized + AsMut<Wtf8>>(x: &mut T) -> &mut Self {
+        x.as_mut()
+    }
+

```

Comment:
```suggestion
```

This will not be useful because impl AsMut<Wtf8> for str is unsound, as I'll show later.